### ISSUE TITLE
[0.F stable] Map generate lamp posts blocking dump gates

### ISSUE THREAD

**Issue by PatrikLundell** [CleverRaven/Cataclysm-DDA#50279]
### Describe the bug

A dump site found had a vehicle in it (behind locked gates). The vehicle was functional, so I picked the gates and drove it out, resulting in a crash as it hit the lamp post on the sidewalk outside.

### Steps To Reproduce

1. Generate a world
2. Find a dump site at a city street

or look at the attacked save

### Expected behavior

Map generation shouldn't generate combinations that conflict.

### Versions and configuration

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: bc28921 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions]
]

### Additional context

The entrance to the dump should probably be at one end of the tile rather than the center. I'd also suggest examining other tiles with exits to verify they don't have those at the center.

**Comment by PatrikLundell** [comment]
Another example of a different clash from the same game (but different town):
![Screenshot (1)](https://user-images.githubusercontent.com/22739822/127573919-07674ce4-4eff-4c47-b656-2934c36931b1.png)
In this case it's possible to pass by to the side of the street lamp, but it still doesn't make real world sense,

**Comment by actual-nh** [comment]
@eltank: I wonder if the obstacle detection code in the automatic-driving PR could be used to help detect situations like this (and others like that one with gates too narrow for farm vehicles or whatever it was) - prior to putting in things like crashed vehicles that could reasonably have happened later, make sure all the places that intact vehicles can spawn actually have a way to get them out and, say, to the nearest road (assuming road or amphibious).

**Comment by eltank** [comment]
While trying to understand/reproduce the bug, I found a pickup truck in a dump site. With careful maneuvering I was able to get it out without crashing into the light post or anything else.
Autodrive will eventually do this careful maneuvering for you (right now autodrive is disabled when you're not standing on a road tile to begin with).
But even if I had to smash some fences to get the vehicle out, I would only consider that a minor inconvenience if I was in desperate need of a vehicle.

> I wonder if the obstacle detection code in the automatic-driving PR could be used to help detect situations like this

Maybe, but that's pretty expensive in terms of CPU cycles and also far from trivial.

![image](https://user-images.githubusercontent.com/8000047/127576264-ac021b3c-a698-403b-bc37-cf69f738b334.png)

**Comment by PatrikLundell** [comment]
If you intend to use the auto driving for testing purposes, that seems like a good use (assuming it can be called at the right time, i.e. before any other vehicles have been spawned, for every vehicle, possibly generating all of them into a "to be placed" list and then "actually" place them later), but it doesn't sound like a good approach for an actual game, as a debugged game shouldn't spawn things incorrectly, and it sounds like a costly thing to do for every new tile.

(I wasn't in a desperate need for a vehicle in my case, although I took two whacks at the lamp post to see if it would work, which it didn't [the lamp posts make city travel a lot harder, since you can no longer freely bypass cars dropped in the road via the sidewalks, so I foresee a need to smash some eventually], but yes, if I had been desperate I would have gotten it out).

**Comment by actual-nh** [comment]
Using it for testing/debugging instead of for "normal" games is an excellent idea.

**Comment by Night-Pryanik** [comment]
This issue was already known at the moment of adding the streetlights to the game, and I'm afraid there is no easy fix for it. Right now the streetlights are placed simply by equal spacing s between them, and if we want to make "smart" placements, we will need to take potential streetlight location into account, checking for existence of various driveways. And that means hardcoded checks for dozens of terrain types (concrete, yellow concrete, sidewalk, pavement, yellow pavement etc).

The more easier way could be to simply rework the problematic locations so their sideways should be slightly shifted from the center of the submap, as streetlights are spawned roughly in the center of the submap (on x axis).

**Comment by PatrikLundell** [comment]
Does the fact that the issue was known also mean the set of problematic locations is known as well, or is there still a value in collecting problematic cases?

I agree it seems like the easier solution would be to tweak each problematic location individually. Not only does it refrain from having the contents of neighboring tiles affect the placement of street lights, but it also avoids screwing up driving for players who may expect street lights to be placed in an equidistant manner, and thus suddenly find lamps blocking their weaving around blockages when the way "should" have been clear (of lamp posts, at least).

**Comment by Night-Pryanik** [comment]
> Does the fact that the issue was known also mean the set of problematic locations is known as well, or is there still a value in collecting problematic cases?

No, the set of such locations isn't known and thus it will be very helpful to create it.

**Comment by PatrikLundell** [comment]
Another one:
![Screenshot (2)](https://user-images.githubusercontent.com/22739822/127765257-0720326b-28f0-42fb-82a5-5754722f4858.png)

**Comment by Procyonae** [comment]
Would replacing sidewalk and lights between the road and pavement/pavement_y post gen with pavement in straight lines perpendicular to the road be an alright fix? It wouldn't fix any instances that don't have pavement obviously and being straight could look quite off but it would at least line up with all existing maps to some degree unlike making a new flag to use in place of SIDEWALK for driveways etc and I can't really think of anything better without getting considerably more complicated

**Comment by Procyonae** [comment]
Actually I think a better still *relatively* simple way would be allowing "existing" true connections for "city_building"'s and adding a new array field that lets you set where you want the from connection to be so like `"connections": [ { "point": [ 0, -1, 0 ], "terrain": "road", "from": [ 0, 0, 0 ], "alignment": [ 3, 8 ], "existing": true } ],` would join it but instead of a regular 16 wide join it would be a 6 wide join from a relative 3 to 8 and probably not use any yellow so you could have custom single lane connections. That would also fix the issue where multitile city buildings don't spawn with the correct rotation bc you'd be able to define it.
